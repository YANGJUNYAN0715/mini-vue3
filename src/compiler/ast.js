export const NodeTypes = {
    ROOT: 'ROOT',
    ELEMENT: 'ELEMENT',
    TEXT: 'TEXT',
    SIMPLE_EXPRESSION: 'SIMPLE_EXPRESSION',
    INTERPOLATION: 'INTERPOLATION',
    ATTRIBUTE: 'ATTRIBUTE',
    DIRECTIVE: 'DIRECTIVE',

    // containers
    COMPOUND_EXPRESSION: 'COMPOUND_EXPRESSION',
    IF: 'IF',
    IF_BRANCH: 'IF_BRANCH',
    FOR: 'FOR',
    TEXT_CALL: 'TEXT_CALL',
    // codegen
    VNODE_CALL: 'VNODE_CALL',
    JS_CALL_EXPRESSION: 'JS_CALL_EXPRESSION',
    JS_OBJECT_EXPRESSION: 'JS_OBJECT_EXPRESSION',
    JS_PROPERTY: 'JS_PROPERTY',


    JS_ARRAY_EXPRESSION: 'JS_ARRAY_EXPRESSION',
    JS_FUNCTION_EXPRESSION: 'JS_FUNCTION_EXPRESSION',
    JS_CONDITIONAL_EXPRESSION: 'JS_CONDITIONAL_EXPRESSION',
    JS_CACHE_EXPRESSION: 'JS_CACHE_EXPRESSION',
}

export const ElementTypes = {
    ELEMENT: 'ELEMENT',
    COMPONENT: 'COMPONENT',
}

export function createRoot(children) {
    return {
        type: NodeTypes.ROOT,
        children,
        helpers: [],
        components: [],
        directives: [],
        hoists: [],
        imports: [],
        cached: 0,
        temps: 0,
        codegenNode: undefined,
    }
}
