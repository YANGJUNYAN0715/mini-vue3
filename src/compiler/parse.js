import { NodeTypes, createRoot, ConstantTypes, ElementTypes } from "./ast";
import { isVoidTag, isNativeTag } from './index'

const TagType = {
  Start: 'Start',
  End: 'End'
};

export function baseParse(content) {
  const context = createParserContext(content)
  return createRoot(
    parseChildren(context, [])
  )
}

function createParserContext(content) {
  return {
    options: {
      delimiters: ['{{', '}}'],
      isVoidTag,
      isNativeTag,
    },
    originalSource: content,
    source: content
  }
}

function parseChildren(context, ancestors) {
  const nodes = []

  while (!isEnd(context, ancestors)) {
    const s = context.source;
    let node;
    if (startsWith(s, context.options.delimiters[0])) {
      // '{{'
      node = parseInterpolation(context)
    } else if (s[0] === '<') {
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context, ancestors)
      }
    }

    if (!node) {
      node = parseText(context)
    }

    if (Array.isArray(node)) {
      for (let i = 0; i < node.length; i++) {
        nodes.push(node[i])
      }
    } else {
      nodes.push(node)
    }
  }

  // Whitespace management for more efficient output
  // (same as v2 whitespace: 'condense')
  let removedWhitespace = false
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    if (node.type === NodeTypes.TEXT) {
      // 正则：全是空格为false，只要有一个其他字符为true
      // 加上!，此条件为：此node节点全是空格
      if (!/[^\t\r\n\f ]/.test(node.content)) {
        const prev = nodes[i - 1]
        const next = nodes[i + 1]
        // 如果空节点处于最后一个或最前一个，则删除
        if (!prev || !next) {
          removedWhitespace = true
          nodes[i] = null
        } else {
          // 否则，用单个空字符替代
          node.content = ' '
        }
      } else {
        // 合并空格
        node.content = node.content.replace(/[\t\r\n\f ]+/g, ' ')
      }
    }
  }
  return removedWhitespace ? nodes.filter(Boolean) : nodes
}

function last(arr) {
  return arr[arr.length - 1]
}

function isEnd(context, ancestors) {
  const s = context.source;
  if (startsWith(s, '</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      if (startsWithEndTagOpen(s, ancestors[i].tag)) {
        return true;
      }
    }
  }
  return !s
}

function startsWithEndTagOpen(source, tag) {
  return (
    startsWith(source, '</') &&
    // tag名相同
    source.substr(2, tag.length).toLowerCase() === tag.toLowerCase() &&
    // tag后以空格或'>'结尾。但不知道为什么要|| '>'
    /[\t\r\n\f />]/.test(source[2 + tag.length] || '>')
  )
}

function startsWith(source, searchString) {
  return source.startsWith(searchString)
}

function parseInterpolation(context) {
  const [open, close] = context.options.delimiters

  advanceBy(context, open.length)
  const closeIndex = context.source.indexOf(close, open.length)

  const content = parseTextData(context, closeIndex).trim()
  advanceBy(context, close.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      isStatic: false,
      // Set `isConstant` to false by default and will decide in transformExpression
      constType: ConstantTypes.NOT_CONSTANT,
      content,
    },
  }
}

function advanceBy(context, numberOfCharacters) {
  const { source } = context
  context.source = source.slice(numberOfCharacters)
}

// 没有trim
function parseTextData(context, length) {
  const rawText = context.source.slice(0, length)
  advanceBy(context, length)
  return rawText
}

function parseText(context) {
  const endTokens = ['<', context.options.delimiters[0]]

  // 寻找text最近的endIndex。因为遇到'<'或'{{'都可能结束
  let endIndex = context.source.length
  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i], 1)
    if (index !== -1 && endIndex > index) {
      endIndex = index
    }
  }

  const content = parseTextData(context, endIndex)

  return {
    type: NodeTypes.TEXT,
    content,
  }
}

function parseElement(context, ancestors) {
  // Start tag.
  const parent = last(ancestors)
  const element = parseTag(context, TagType.Start, parent)

  if (element.isSelfClosing || context.options.isVoidTag(element.tag)) {
    return element
  }

  // Children.
  ancestors.push(element)
  const children = parseChildren(context, ancestors)
  ancestors.pop()

  element.children = children

  // End tag.
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End, parent)
  }

  return element
}

function parseTag(context, type, parent) {
  // Tag open.
  const match = /^<\/?([a-z][^\t\r\n\f />]*)/i.exec(context.source)
  const tag = match[1]

  advanceBy(context, match[0].length)
  advanceSpaces(context)

  // Attributes.
  let props = parseAttributes(context, type)

  // Tag close.
  let isSelfClosing = startsWith(context.source, '/>')

  advanceBy(context, isSelfClosing ? 2 : 1)

  let tagType = ElementTypes.ELEMENT
  if (isComponent(tag, context)) {
    tagType = ElementTypes.COMPONENT
  }

  return {
    type: NodeTypes.ELEMENT,
    tag,
    tagType,
    props,
    isSelfClosing,
    children: [],
    codegenNode: undefined // to be created during transform phase
  }
}

function isComponent(tag, context) {
  const { options } = context;
  // todo isNativeTag
  if (options.isNativeTag) {
    if (!options.isNativeTag(tag)) {
      return true;
    }
  }
  return false;
}

function advanceSpaces(context) {
  const match = /^[\t\r\n\f ]+/.exec(context.source)
  if (match) {
    advanceBy(context, match[0].length)
  }
}

function parseAttributes(context, type) {
  const props = []
  const attributeNames = new Set()
  while (
    context.source.length > 0 &&
    !startsWith(context.source, '>') &&
    !startsWith(context.source, '/>')
  ) {
    const attr = parseAttribute(context, attributeNames)
    if (type === TagType.Start) {
      props.push(attr)
    }
    advanceSpaces(context)
  }
  return props
}

function parseAttribute(context, nameSet) {
  // Name.
  // todo patt
  const match = /^[^\t\r\n\f />][^\t\r\n\f />=]*/.exec(context.source)
  const name = match[0]

  nameSet.add(name)

  advanceBy(context, name.length)

  // Value
  let value

  if (/^[\t\r\n\f ]*=/.test(context.source)) {
    advanceSpaces(context)
    advanceBy(context, 1)
    advanceSpaces(context)
    value = parseAttributeValue(context)
  }


  if (!context.inVPre && /^(v-|:|@|#)/.test(name)) {
    const match = /(?:^v-([a-z0-9-]+))?(?:(?::|^@|^#)(\[[^\]]+\]|[^\.]+))?(.+)?$/i.exec(
      name
    )

    let dirName =
      match[1] ||
      (startsWith(name, ':') ? 'bind' : startsWith(name, '@') ? 'on' : 'slot')
    let arg
    if (match[2]) {
      let content = match[2]
      let isStatic = true

      if (content.startsWith('[')) {
        isStatic = false
        content = content.substr(1, content.length - 2)
      }

      arg = {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content,
        isStatic,
        constType: isStatic
          ? ConstantTypes.CAN_STRINGIFY
          : ConstantTypes.NOT_CONSTANT,
      }
    }

    return {
      type: NodeTypes.DIRECTIVE,
      name: dirName,
      exp: value && {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: value.content,
        isStatic: false,
        // Treat as non-constant by default. This can be potentially set to
        // other values by `transformExpression` to make it eligible for hoisting.
        constType: ConstantTypes.NOT_CONSTANT,
      },
      arg,
    }
  }

  return {
    type: NodeTypes.ATTRIBUTE,
    name,
    value: value && {
      type: NodeTypes.TEXT,
      content: value.content,
    },
  }
}

function parseAttributeValue(context) {
  let content

  const quote = context.source[0]
  // 不考虑没有引号的情况
  advanceBy(context, 1)

  const endIndex = context.source.indexOf(quote)
  content = parseTextData(context, endIndex)
  advanceBy(context, 1)

  return { content }
}
