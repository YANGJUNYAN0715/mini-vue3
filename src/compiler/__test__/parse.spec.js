import { baseParse } from '../parse'
import { NodeTypes, ElementTypes } from '../ast'

describe('compiler: parse', () => {
    describe('Text', () => {
        test('simple text', () => {
            const ast = baseParse('some text')
            const text = ast.children[0]

            expect(text).toStrictEqual({
                type: NodeTypes.TEXT,
                content: 'some text',
            })
        })

        test('text with interpolation', () => {
            const ast = baseParse('some {{ foo + bar }} text')
            const text1 = ast.children[0]
            const text2 = ast.children[2]

            expect(text1).toStrictEqual({
                type: NodeTypes.TEXT,
                content: 'some ',
            })
            expect(text2).toStrictEqual({
                type: NodeTypes.TEXT,
                content: ' text',
            })
        })

        test('text with interpolation which has `<`', () => {
            const ast = baseParse('some {{ a<b && c>d }} text')
            const text1 = ast.children[0]
            const text2 = ast.children[2]

            expect(text1).toStrictEqual({
                type: NodeTypes.TEXT,
                content: 'some ',
            })
            expect(text2).toStrictEqual({
                type: NodeTypes.TEXT,
                content: ' text',
            })
        })

        test('text with interpolation which has `<`', () => {
            const ast = baseParse('some {{ a<b && c>d }} text')
            const text1 = ast.children[0]
            const text2 = ast.children[2]

            expect(text1).toStrictEqual({
                type: NodeTypes.TEXT,
                content: 'some ',
            })
            expect(text2).toStrictEqual({
                type: NodeTypes.TEXT,
                content: ' text',
            })
        })

        test('text with mix of tags and interpolations', () => {
            const ast = baseParse('some <span>{{ foo < bar + foo }} text</span>')
            const text1 = ast.children[0]
            const text2 = ast.children[1].children[1]

            expect(text1).toStrictEqual({
                type: NodeTypes.TEXT,
                content: 'some ',
            })
            expect(text2).toStrictEqual({
                type: NodeTypes.TEXT,
                content: ' text',
            })
        })
    })
    describe('Interpolation', () => {
        test('simple interpolation', () => {
            const ast = baseParse('{{message}}')
            const interpolation = ast.children[0]

            expect(interpolation).toStrictEqual({
                type: NodeTypes.INTERPOLATION,
                content: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: `message`,
                    isStatic: false,
                },
            })
        })

        test('it can have tag-like notation', () => {
            const ast = baseParse('{{ a<b }}')
            const interpolation = ast.children[0]

            expect(interpolation).toStrictEqual({
                type: NodeTypes.INTERPOLATION,
                content: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: `a<b`,
                    isStatic: false,
                },
            })
        })

        test('it can have tag-like notation (2)', () => {
            const ast = baseParse('{{ a<b }}{{ c>d }}')
            const interpolation1 = ast.children[0]
            const interpolation2 = ast.children[1]

            expect(interpolation1).toStrictEqual({
                type: NodeTypes.INTERPOLATION,
                content: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: `a<b`,
                    isStatic: false,
                },
            })

            expect(interpolation2).toStrictEqual({
                type: NodeTypes.INTERPOLATION,
                content: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    isStatic: false,
                    content: 'c>d',
                },
            })
        })

        test('it can have tag-like notation (3)', () => {
            const ast = baseParse('<div>{{ "</div>" }}</div>')
            const element = ast.children[0]
            const interpolation = element.children[0]

            expect(interpolation).toStrictEqual({
                type: NodeTypes.INTERPOLATION,
                content: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    isStatic: false,
                    content: '"</div>"',
                },
            })
        })

        test('delimiters', () => {
            const ast = baseParse('<p>{{msg}}</p>')
            const element = ast.children[0]
            const interpolation = element.children[0]

            expect(interpolation).toStrictEqual({
                type: NodeTypes.INTERPOLATION,
                content: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: `msg`,
                    isStatic: false,
                }
            })
        })
    })

    describe('Element', () => {
        test('simple div', () => {
            const ast = baseParse('<div>hello</div>')
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [],
                isSelfClosing: false,
                children: [
                    {
                        type: NodeTypes.TEXT,
                        content: 'hello',
                    }
                ],
            })
        })

        test('empty', () => {
            const ast = baseParse('<div></div>')
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [],
                isSelfClosing: false,
                children: [],
            })
        })

        test('self closing', () => {
            const ast = baseParse('<div/>after')
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [],

                isSelfClosing: true,
                children: [],
            })
        })

        test('void element', () => {
            const ast = baseParse('<img>after')
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'img',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [],

                isSelfClosing: false,
                children: [],
            })
        })

        test('native element with `isNativeTag`', () => {
            const ast = baseParse('<div></div><comp></comp><Comp></Comp>')

            expect(ast.children[0]).toMatchObject({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT
            })

            expect(ast.children[1]).toMatchObject({
                type: NodeTypes.ELEMENT,
                tag: 'comp',
                tagType: ElementTypes.COMPONENT
            })

            expect(ast.children[2]).toMatchObject({
                type: NodeTypes.ELEMENT,
                tag: 'Comp',
                tagType: ElementTypes.COMPONENT
            })
        })

        test('attribute with no value', () => {
            const ast = baseParse('<div id></div>')
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'id',
                        value: undefined,
                    }
                ],

                isSelfClosing: false,
                children: [],
            })
        })

        test('attribute with empty value, double quote', () => {
            const ast = baseParse('<div id=""></div>')
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'id',
                        value: {
                            type: NodeTypes.TEXT,
                            content: '',
                        },
                    }
                ],

                isSelfClosing: false,
                children: [],
            })
        })

        test('attribute with empty value, single quote', () => {
            const ast = baseParse("<div id=''></div>")
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'id',
                        value: {
                            type: NodeTypes.TEXT,
                            content: '',
                        },
                    }
                ],

                isSelfClosing: false,
                children: [],
            })
        })

        test('attribute with value, double quote', () => {
            const ast = baseParse('<div id=">\'"></div>')
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'id',
                        value: {
                            type: NodeTypes.TEXT,
                            content: ">'",
                        },
                    }
                ],

                isSelfClosing: false,
                children: [],
            })
        })

        test('attribute with value, single quote', () => {
            const ast = baseParse("<div id='>\"'></div>")
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'id',
                        value: {
                            type: NodeTypes.TEXT,
                            content: '>"',
                        },
                    }
                ],

                isSelfClosing: false,
                children: [],
            })
        })

        test('multiple attributes', () => {
            const ast = baseParse('<div id="a" class="c" inert style=\'\'></div>')
            const element = ast.children[0]

            expect(element).toStrictEqual({
                type: NodeTypes.ELEMENT,
                tag: 'div',
                tagType: ElementTypes.ELEMENT,
                codegenNode: undefined,
                props: [
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'id',
                        value: {
                            type: NodeTypes.TEXT,
                            content: 'a',
                        },
                    },
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'class',
                        value: {
                            type: NodeTypes.TEXT,
                            content: 'c',
                        },
                    },
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'inert',
                        value: undefined,
                    },
                    {
                        type: NodeTypes.ATTRIBUTE,
                        name: 'style',
                        value: {
                            type: NodeTypes.TEXT,
                            content: '',
                        },
                    }
                ],

                isSelfClosing: false,
                children: [],
            })
        })

        test('directive with no value', () => {
            const ast = baseParse('<div v-if/>')
            const directive = (ast.children[0]).props[0]

            expect(directive).toStrictEqual({
                type: NodeTypes.DIRECTIVE,
                name: 'if',
                arg: undefined,
                exp: undefined,
            })
        })

        test('directive with value', () => {
            const ast = baseParse('<div v-if="a"/>')
            const directive = (ast.children[0]).props[0]

            expect(directive).toStrictEqual({
                type: NodeTypes.DIRECTIVE,
                name: 'if',
                arg: undefined,
                exp: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: 'a',
                    isStatic: false,
                },
            })
        })

        test('directive with argument', () => {
            const ast = baseParse('<div v-on:click/>')
            const directive = (ast.children[0]).props[0]

            expect(directive).toStrictEqual({
                type: NodeTypes.DIRECTIVE,
                name: 'on',
                arg: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: 'click',
                    isStatic: true,
                },
                exp: undefined,
            })
        })

        test('v-bind shorthand', () => {
            const ast = baseParse('<div :a="b" />')
            const directive = (ast.children[0]).props[0]

            expect(directive).toStrictEqual({
                type: NodeTypes.DIRECTIVE,
                name: 'bind',
                arg: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: 'a',
                    isStatic: true,
                },
                exp: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: 'b',
                    isStatic: false,
                },
            })
        })

        test('v-on shorthand', () => {
            const ast = baseParse('<div @a="b" />')
            const directive = (ast.children[0]).props[0]

            expect(directive).toStrictEqual({
                type: NodeTypes.DIRECTIVE,
                name: 'on',
                arg: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: 'a',
                    isStatic: true,
                },
                exp: {
                    type: NodeTypes.SIMPLE_EXPRESSION,
                    content: 'b',
                    isStatic: false,
                },
            })
        })

        test('end tags are case-insensitive.', () => {
            const ast = baseParse('<div>hello</DIV>after')
            const element = ast.children[0]
            const text = element.children[0]

            expect(text).toStrictEqual({
                type: NodeTypes.TEXT,
                content: 'hello',
            })
        })

        test('self closing single tag', () => {
            const ast = baseParse('<div :class="{ some: condition }" />')

            expect(ast.children).toHaveLength(1)
            expect(ast.children[0]).toMatchObject({ tag: 'div' })
        })

        test('self closing multiple tag', () => {
            const ast = baseParse(
                `<div :class="{ some: condition }" />\n` +
                `<p v-bind:style="{ color: 'red' }"/>`
            )

            expect(ast).toMatchSnapshot()

            expect(ast.children).toHaveLength(3)
            expect(ast.children[0]).toMatchObject({ tag: 'div' })
            expect(ast.children[2]).toMatchObject({ tag: 'p' })
        })

        test('valid html', () => {
            const ast = baseParse(
                `<div :class="{ some: condition }">\n` +
                `  <p v-bind:style="{ color: 'red' }"/>\n` +
                `</div>`
            )

            expect(ast).toMatchSnapshot()

            expect(ast.children).toHaveLength(1)
            const el = ast.children[0]
            expect(el).toMatchObject({
                tag: 'div'
            })
            expect(el.children).toHaveLength(3)
            expect(el.children[1]).toMatchObject({
                tag: 'p'
            })
        })
    })

    describe('whitespace management when adopting strategy condense', () => {
        it('should remove whitespaces w/ newline between elements', () => {
            const ast = baseParse(`<div/> \n <div/> \n <div/>`)
            expect(ast.children.length).toBe(5)
            expect(ast.children[1].content).toBe(' ')
        })

        it('should NOT remove whitespaces w/ newline between interpolations', () => {
            const ast = baseParse(`{{ foo }} \n {{ bar }}`)
            expect(ast.children.length).toBe(3)
            expect(ast.children[0].type).toBe(NodeTypes.INTERPOLATION)
            expect(ast.children[1]).toMatchObject({
                type: NodeTypes.TEXT,
                content: ' '
            })
            expect(ast.children[2].type).toBe(NodeTypes.INTERPOLATION)
        })

        it('should NOT remove whitespaces w/o newline between elements', () => {
            const ast = baseParse(`<div/> <div/> <div/>`)
            expect(ast.children.length).toBe(5)
            expect(ast.children.map(c => c.type)).toMatchObject([
                NodeTypes.ELEMENT,
                NodeTypes.TEXT,
                NodeTypes.ELEMENT,
                NodeTypes.TEXT,
                NodeTypes.ELEMENT
            ])
        })

        it('should condense consecutive whitespaces in text', () => {
            const ast = baseParse(`   foo  \n    bar     baz     `)
            expect((ast.children[0]).content).toBe(` foo bar baz `)
        })
    })
})
