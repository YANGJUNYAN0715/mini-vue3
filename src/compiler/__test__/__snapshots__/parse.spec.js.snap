// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: parse Element self closing multiple tag 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "directives": Array [
        Object {
          "arg": Object {
            "content": "class",
            "isStatic": true,
            "type": "SIMPLE_EXPRESSION",
          },
          "exp": Object {
            "content": "{ some: condition }",
            "isStatic": false,
            "type": "SIMPLE_EXPRESSION",
          },
          "name": "bind",
          "type": "DIRECTIVE",
        },
      ],
      "isSelfClosing": true,
      "props": Array [],
      "tag": "div",
      "tagType": "ELEMENT",
      "type": "ELEMENT",
    },
    Object {
      "children": Array [],
      "directives": Array [
        Object {
          "arg": Object {
            "content": "style",
            "isStatic": true,
            "type": "SIMPLE_EXPRESSION",
          },
          "exp": Object {
            "content": "{ color: 'red' }",
            "isStatic": false,
            "type": "SIMPLE_EXPRESSION",
          },
          "name": "bind",
          "type": "DIRECTIVE",
        },
      ],
      "isSelfClosing": true,
      "props": Array [],
      "tag": "p",
      "tagType": "ELEMENT",
      "type": "ELEMENT",
    },
  ],
  "type": "ROOT",
}
`;

exports[`compiler: parse Element valid html 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "directives": Array [
            Object {
              "arg": Object {
                "content": "style",
                "isStatic": true,
                "type": "SIMPLE_EXPRESSION",
              },
              "exp": Object {
                "content": "{ color: 'red' }",
                "isStatic": false,
                "type": "SIMPLE_EXPRESSION",
              },
              "name": "bind",
              "type": "DIRECTIVE",
            },
          ],
          "isSelfClosing": true,
          "props": Array [],
          "tag": "p",
          "tagType": "ELEMENT",
          "type": "ELEMENT",
        },
      ],
      "directives": Array [
        Object {
          "arg": Object {
            "content": "class",
            "isStatic": true,
            "type": "SIMPLE_EXPRESSION",
          },
          "exp": Object {
            "content": "{ some: condition }",
            "isStatic": false,
            "type": "SIMPLE_EXPRESSION",
          },
          "name": "bind",
          "type": "DIRECTIVE",
        },
      ],
      "isSelfClosing": false,
      "props": Array [],
      "tag": "div",
      "tagType": "ELEMENT",
      "type": "ELEMENT",
    },
  ],
  "type": "ROOT",
}
`;
